Update for 2025.3

sh41 opened the pull request:
Hi, 

I saw the chat in #3711 about the difficulties with the changes that JB have made to the APIs but I wanted to carry on using this excellent plugin with 2025.3, so I've created a version that only works on IntelliJ IDEA 2025.3 Beta. It's based off all the good work that @joshuataylor has done in #3711, but I haven't done much real world testing with it yet. 

I've got the tests working (at least on my local setup) and have resolved a bunch of warnings that I thought were causing issues, but actually weren't. 

The main reason I created this was to resolve an issue I was having opening Elixir projects for the first time in IDEA 2025.3 Beta. It would crash with the error below (reproduced here to help others with the same problem find it). 
````   
   Internal error
   
   com.intellij.openapi.diagnostic.RuntimeExceptionWithAttachments: Access is allowed from Event Dispatch Thread (EDT) only; If you access or modify model on EDT consider wrapping your code in WriteIntentReadAction  or ReadAction; see https://jb.gg/ij-platform-threading for details
   Current thread: Thread[#93,DefaultDispatcher-worker-42,5,main] 177714018 (EventQueue.isDispatchThread()=false)
   SystemEventQueueThread: Thread[#42,AWT-EventQueue-0,6,main] 993174760
           at com.intellij.util.concurrency.ThreadingAssertions.createThreadAccessException(ThreadingAssertions.java:257)
           at com.intellij.util.concurrency.ThreadingAssertions.throwThreadAccessException(ThreadingAssertions.java:248)
           at com.intellij.util.concurrency.ThreadingAssertions.assertEventDispatchThread(ThreadingAssertions.java:89)
           at com.intellij.ide.impl.DataManagerImpl.getDataContext(DataManagerImpl.kt:198)
           at com.intellij.ide.impl.DataManagerImpl.getDataContext(DataManagerImpl.kt:231)
           at com.intellij.projectImport.ProjectImportBuilder.getCurrentProject(ProjectImportBuilder.java:64)
           at org.elixir_lang.mix.project._import.Builder.setProjectRoot(Builder.kt:162)
           at org.elixir_lang.mix.project.OpenProcessor.doQuickImport(OpenProcessor.kt:20)
           at com.intellij.projectImport.ProjectOpenProcessorBase.openProjectAsync$suspendImpl(ProjectOpenProcessorBase.kt:97)
           at com.intellij.projectImport.ProjectOpenProcessorBase.openProjectAsync(ProjectOpenProcessorBase.kt)
           at com.intellij.ide.impl.ProjectUtil.chooseProcessorAndOpenAsync(ProjectUtil.kt:319)
           at com.intellij.ide.impl.ProjectUtil.openOrImportAsync(ProjectUtil.kt:246)
           at com.intellij.ide.impl.ProjectUtil$openOrImportAsync$1.invokeSuspend(ProjectUtil.kt)
           at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:34)
           at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:100)
           at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:610)
           at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runDefaultDispatcherTask(CoroutineScheduler.kt:1194)
           at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:906)
           at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:775)
           at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:762)``` 
````
Hope this is useful to someone!

[intellij-elixir-22.0.0-pre+2025.3-20251208134427.zip](https://github.com/user-attachments/files/24041895/intellij-elixir-22.0.0-pre%2B2025.3-20251208134427.zip) - built with IDEA Ultimate 2025.3 libraries.

sh41 committed:
Extract JBR setup into a reusable composite action

Ensures that the same JBR is used in all actions.

(cherry picked from commit 27e02e93e3a81bc3955568f6d0baf9734b5b97f9)

# Conflicts:
#	.github/workflows/release.yml
#	.github/workflows/test.yml

joshuataylor committed:
Update to release setup-java@v5

(cherry picked from commit e1332fe7893c82e5beae25dd283b6b10a0e3a6eb)

joshuataylor committed:
Add gradle cache to github actions

(cherry picked from commit 5f3ad926cea4f32b1aec5371ac45d1159a74d1f3)

joshuataylor committed:
Make canary release a manual process

(cherry picked from commit 07d95426cdbd622b6bb7e03cf9d06206f86d94fa)

joshuataylor committed:
Bump checkout to v5

(cherry picked from commit 9e4e921368fd7d9b23eea8a34e7d04ba9194941c)

joshuataylor committed:
Add workflow_dispatch to Github actions test

(cherry picked from commit 1e3b5f97c0d4e2b4309b33fe8a6e91fb834f47dc)

joshuataylor committed:
Bump to actions/checkout@v5

(cherry picked from commit c6de7f2f2e665f05bedd2a38bd8fe8d872350291)

sh41 committed:
Upgrade to IntelliJ Platform 2025.3+ and enable EAP versions

- Switch from IC to IU (IntelliJ IDEA Community to Ultimate)
- Update platform versions to 2025.3-EAP-SNAPSHOT
- Add dynamic EAP version fetching from JetBrains API
- Replace intelliLang plugin with intellij.platform.langInjection module
- Update Gradle plugin from 2.7.2 to 2.10.4

sh41 committed:
Ensure full teardown in PlatformTestCase.kt

sh41 committed:
Bump dependencies

sh41 committed:
Remove unused dependencies

sh41 committed:
Update removed `StringUtil.createToStringFunction`

sh41 committed:
Replace deprecated Char.toInt() with Char.code

The Char.toInt() method has been deprecated in Kotlin stdlib.
Replace all occurrences with the Char.code property.

sh41 committed:
Replace deprecated String capitalize/decapitalize

The String.capitalize() and String.decapitalize() methods have been
deprecated in Kotlin stdlib. Replace with replaceFirstChar { it.uppercase() }
and replaceFirstChar { it.lowercase() } respectively.

sh41 committed:
Replace deprecated StandardCharsets.UTF_8

The java.nio.charset.StandardCharsets.UTF_8 static field has been
deprecated. Replace with Charset.forName("UTF-8").

sh41 committed:
Update addBrowseFolderListener to use new API

The 4-parameter addBrowseFolderListener(title, description, project, descriptor)
has been deprecated. Use the 2-parameter version and move title/description
into the FileChooserDescriptor via withTitle() and withDescription().

sh41 committed:
Replace deprecated PsiElement offset extensions

The startOffset and endOffset extension properties from
com.intellij.refactoring.suggested are deprecated. Replace with
direct access to PsiElement.textRange.startOffset and
PsiElement.textRange.endOffset.

sh41 committed:
Fix FindUsages

2025.3 reverts to the same behaviour as 2021.3 which caused bug #2374 resolved in #2461. This just includes the 2025.3 build in the workaround logic.

sh41 committed:
Fix FindUsagesTest - Switch to `testFindUsages` rather than `testFindUsagesAction`

This avoids a "Double smart pointer removal" exception in the teardown while retaining the ability to check the actual results of the usage search

sh41 committed:
Implement deferred project root scanning in Mix import builder

The scan is now performed lazily when `getList()` is called, rather than immediately in `setProjectRoot()`. Fixes IntelliJ 2025.3 compatibility by returning false from OpenProcessor.

sh41 committed:
Add Latest 2025.3 Beta version to test matrix & remove all other versions

This won't work on previous versions. Also, disable the dynamic EAP version because that's only intended for dev work.

sh41 committed:
Resolve "Unable to load JUnit4 runner to calculate Ignored test cases" warning

sh41 committed:
Resolve "Not annotated parameter overrides @NotNull parameter" warning

sh41 committed:
Simplify assertions

sh41 committed:
Simplify type definitions

sh41 committed:
Remove unused deps

sh41 committed:
Exclude abstract test classes rather than @Ignore them.

The base classes are not tests, but they were showing as "Ignored". This excludes them from the search pattern so that they are not mentioned as Ignored in the completed test report.

sh41 committed:
Resolve linter warnings

sh41 committed:
Resolve Gradle deprecation warnings:

```
Build file '/dev/build.gradle': line 112
The org.gradle.api.plugins.JavaPluginConvention type has been deprecated. This is scheduled to be removed in Gradle 9.0. Consult the upgrading guide for further information: https://docs.gradle.org/8.14.3/userguide/upgrading_version_8.html#java_convention_deprecation
        at build_8l04vxwyktujcq3n0cbxr7t3w$_run_closure2.doCall$original(/dev/build.gradle:112)
        (Run with --stacktrace to get the full stack trace of this deprecation warning.)
        at build_8l04vxwyktujcq3n0cbxr7t3w.run(/dev/build.gradle:110)
        (Run with --stacktrace to get the full stack trace of this deprecation warning.)
Build file '/dev/build.gradle': line 113
The org.gradle.api.plugins.JavaPluginConvention type has been deprecated. This is scheduled to be removed in Gradle 9.0. Consult the upgrading guide for further information: https://docs.gradle.org/8.14.3/userguide/upgrading_version_8.html#java_convention_deprecation
        at build_8l04vxwyktujcq3n0cbxr7t3w$_run_closure2.doCall$original(/dev/build.gradle:113)
        (Run with --stacktrace to get the full stack trace of this deprecation warning.)
        at build_8l04vxwyktujcq3n0cbxr7t3w.run(/dev/build.gradle:110)
        (Run with --stacktrace to get the full stack trace of this deprecation warning.)
```

sh41 committed:
Resolve `Cannot assign 'GString' to 'GStringImpl` warnings

sh41 committed:
Migrate to lazy task configuration

sh41 committed:
Migrate deprecated `exec` closures to gradle 9.

sh41 committed:
Migrate build.gradle & settings.gradle to kotlin

sh41 committed:
Tidy up jps-builder tests.

sh41 committed:
Remove unused `testCompilation` task

sh41 committed:
Allow tests to access testData

sh41 committed:
Migrate JPS `build.gradle` to kotlin

sh41 committed:
Refactor GHA workflows to share more code and report on test results.

sh41 committed:
Consolidate duplicated config from subprojects and root into allprojects

sh41 commented:
@joshuataylor I spent some more time on this dealing with some of the deprecations and have ended up migrating the `build.gradle` scripts to kotlin and refactoring the GH Actions to be a bit DRYer. I also added in some test reporting which you can see an example of in https://github.com/KronicDeth/intellij-elixir/actions/runs/19750620486?pr=3717 

Now I need to spend some time actually using the plugin! I've uploaded the latest plugin artifact to the summary above if anyone else wants to try it.

sh41 committed:
Bump to 2025.3 Release Candidate

CrystalJewell commented:
@sh41 
Just installed this updated version in the 2025.3 RC `Build #IU-253.28294.251` and right off the bat the first thing I noticed is this is the first time in what feels like years that I haven't been greeted with the report error box as soon as I open a project.  I didn't realize how much of a habit it had become to just close out that popup until I didn't need to.  I almost exclusively work in Elixir so I'd be more than happy to help with testing updated versions.

sh41 committed:
Add `warn_if_outdated` to ignored mix.exs/deps options.

joshuataylor commented:
Legend!

Sorry for the delay on this, I've been busy with other work.

I've also been working on "Elixir Helper", which is a LSP integration of Elixir.

I'll try and have a look at this this week, otherwise it'll be over the weekend.

(Got IRL family stuff going on this week, apologies for the delay)

sh41 commented:
Hi @CrystalJewell, funnily enough I just had the same error pop up and have fixed it. New version now available in the first comment above. 

@joshuataylor, great news on the LSP work. Not sure what is best to do with this version. I don't use RubyMine or anything like that and haven't even looked at what would be needed for it to work on those.

sh41 committed:
Migrate to IntelliJ Platform 2025.3 unified distribution

 - Remove IntellijIdeaCommunity from runIdePlatforms (no longer available in 2025.3+)
 - Use Action wrapper for intellijPlatformTesting.runIde.register() to fix type inference
 - Clean up deprecated Community Edition version properties

sh41 committed:
Migrate to IntelliJ Platform 2025.3 unified distribution

 - Remove IntellijIdeaCommunity from runIdePlatforms (no longer available in 2025.3+)
 - Use Action wrapper for intellijPlatformTesting.runIde.register() to fix type inference
 - Clean up deprecated Community Edition version properties

sh41 committed:
Free up space on the shared runner to allow downloading IDEs for verifyPlugin

joshuataylor commented:
(commented with wrong account)
Yeah I don't use RubyMine either, IMHO we should just merge this branch some point, once you're happy with it, given 2025.3 is in RC now.

Let me know once you're happy with the changes, and I'll review and merge.

Again, absolute legend 👍

sh41 committed:
feat: Support 2025.3 and fix crashes in Small IDEs

- Targeted 2025.3 RC in build.gradle.kts
- Moved Java-dependent extensions to rich-platform-plugin.xml
- Implemented Service pattern for BeforeRunTaskProvider to abstract 'Make' step
- Updated Run Configuration Factories to use safe services

sh41 committed:
refactor: safe module loading in Run Configuration Editors

- Created EditorHelper to centralize module selection logic
- Replaced direct ElixirModuleType usage with String ID lookup
- Prevents NoClassDefFoundError in WebStorm when opening Run settings

sh41 committed:
Refactor build scripts to enable caching

- Add Version Catalog (libs.versions.toml) for dependency management
- Moved EAP version fetching logic to buildSrc/VersionFetcher.kt
- Refactored build.gradle.kts to use declarative plugins and properties
- Fixed test environment race conditions by polling Quoter daemon status
- Cleaned up gradle.properties

sh41 committed:
Restore EOLtoWhitespace test

Was removed in e71b247aacaff389358ca5441c32512cdafd6af9, but seems to run fine, so added back in.

sh41 committed:
chore: upgrade Gradle from 8.14.3 to 9.2.1

sh41 committed:
Add extra debug info for ChooseByNameContributor

sh41 committed:
Resolve `application library named XXX already exists` exception

java.lang.IllegalStateException: application library named Erlang for Elixir 24 already exists
	at com.intellij.workspaceModel.ide.impl.legacyBridge.library.GlobalLibraryTableDelegate.createLibrary$intellij_platform_projectModel_impl(GlobalLibraryTableDelegate.kt:155)
	at com.intellij.workspaceModel.ide.impl.legacyBridge.library.GlobalLibraryTableBridgeImpl.createLibrary(GlobalLibraryTableBridgeImpl.kt:61)
	at org.elixir_lang.facet.sdks.Configurable$addListeners$listener$1.sdkAdded$lambda$1$0(Configurable.kt:126)

sh41 committed:
Resolve `Write access is allowed inside write-action only` exception

com.intellij.openapi.diagnostic.RuntimeExceptionWithAttachments: Write access is allowed inside write-action only (see Application.runWriteAction()); If you access or modify model on EDT consider wrapping your code in WriteIntentReadAction ; see https://jb.gg/ij-platform-threading for details
Current thread: Thread[#46,AWT-EventQueue-0,6,main] 1321404175 (EventQueue.isDispatchThread()=true)
SystemEventQueueThread: (same)
	at com.intellij.util.concurrency.ThreadingAssertions.createThreadAccessException(ThreadingAssertions.java:257)
	at com.intellij.util.concurrency.ThreadingAssertions.throwThreadAccessException(ThreadingAssertions.java:248)
	at com.intellij.util.concurrency.ThreadingAssertions.assertWriteAccess(ThreadingAssertions.java:233)
	at com.intellij.openapi.projectRoots.impl.ProjectJdkImpl.commitChanges(ProjectJdkImpl.java:181)
	at org.elixir_lang.facet.sdk.Editor.reset(Editor.kt:201)
	at org.elixir_lang.facet.sdk.Editor.<init>(Editor.kt:65)
	at org.elixir_lang.facet.sdks.Configurable.updateSdkPanel$lambda$0$0(Configurable.kt:197)
	at org.elixir_lang.facet.sdks.Configurable.updateSdkPanel$lambda$0$1(Configurable.kt:197)

sh41 committed:
Resolve `Please make class org.elixir_lang.beam.file_editor.Provider dumb-aware`

sh41 committed:
Create "mix.exs" constant to remove reliance on PackageManager object.

sh41 committed:
Fix BEAM Stub Indexing and Platform compatibility

- Refactor `ModuleStubElementTypes` to an interface with top-level Kotlin classes to satisfy `StubElementTypeHolderEP` requirements.
- Update `ModuleStubElementType` to use the correct `BEAM` Language instance.
- Align External IDs with field names (UPPER_CASE) to enable the `externalIdPrefix` optimization.
- Increment `STUB_VERSION` to force re-indexing of incompatible data.

sh41 committed:
Upload verifyPlugin reports

CrystalJewell commented:
> @CrystalJewell if you have time to test and provide feedback that would be great.

@sh41 Its been a busy week for me already so the last version of the plugin has def been put to the test with nothing that has errored since the `warn_if_outdated` issue.  I gotchu on the new one 🫶 just downloaded the file and got it installed, will report back with any issues as I use it this week.  Thank you so much for all your work on this!

sh41 committed:
Bump platform to full release 2025.3

sh41 committed:
Fix `Don't know how to walk unquoted variable parent`

Triggered when indexing https://github.com/rrrene/credo/blob/1f5ae9e2dd33210a2a13d3e3de8118f4e34edf9b/lib/credo/check.ex

sh41 committed:
Adjust nameSet size logging thresholds for protocol implementations

Increases thresholds to accommodate real-world protocol implementation counts:
- No logging for ≤30 implementations (was ≥20 error)
- Info logging for 31-60 implementations
- Warn logging for 61-100 implementations
- Error logging for >100 implementations

Accommodates OpenApiSpex's 25 Extendable protocol implementations without spurious error logs while still detecting potentially corrupt stub indices.

joshuataylor commented:
(I'm going to try and merge this now)
